scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala

```diff
@@ -99,6 +118,61 @@ package object hdfs {
         .map(_.getKey.datum())
     }
 
+    /**
+     * Hadoop source flavoured distributed cache - any source supported by Hadoop Filesystem
+     * abstraction can be used as a source for distributed cache.
+     *
+     * @param path to the artifact - it can be local/HDFS/S3/GCS ...
+     * @param conf optional custom Hadoop configuration
+     * @param username optional Hadoop Simple Authentication remote username
+     */
+    def hadoopDistCache[F](path: String,
+                           conf: Configuration = null,
+                           username: String = null)
+                          (initFn: File => F): DistCache[F] = self.pipelineOp {
+      if (self.isTest) {
+        //TODO: if test - can the path be remote HDFS/S3 etc?
+        self.distCache(path)(initFn)
+      } else {
+        require(self.options.getStagingLocation != null,
+          "Staging directory not set - use `--stagingLocation`!")
+        require(path != null, "Artifact path can't be null")
+
+        logger.debug(s"Add '$path' to dist cache")
+
+        val _conf = Option(conf).getOrElse(new Configuration())
+        val target = new Path(self.options.getStagingLocation, path.split("/").last)
```

Comment:
Good idea about extra `dir`. initially I was thinking we could just overwrite files, but hash is probably better.